### PR TITLE
[node-fetch] Add `Fetch` type representing just the `fetch` function

### PR DESCRIPTION
**Pull request by trevor-scheer**
We want to provide a fetcher _interface_ in order to allow users to bring their own fetcher. The typings from `node-fetch` are really close to allowing that, but the `isRedirect` is problematic. This allows us to specify the function type without bringing anything extra.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Review by glasser — APPROVED** [review]
I think this is a good change. I know of software such as Apollo Server that allows you to configure its HTTP fetching by passing in a `fetch`-compatible function as an option. It would be nice to be able to declare these options as `typeof fetch` from `node-fetch`, but then it is awkward that `fetch` has a random `isRedirect` function on it. This explicit function type solves that issue. Thanks!

**Review by nikcorg — CHANGES_REQUESTED** [review]
(Ignore this comment, I'm unable to delete it.)

**Review by nikcorg — CHANGES_REQUESTED** [review]
Are you submitting this change because you want to reuse node-fetch types for your "bring your own fetch" type definition, or because node-fetch types are in conflict with the canonical type definition for fetch (upon which you're basing your own interface)?

(Not really asking for changes, but I want to block a merge before this point is clarified.)

**Review by AndrewLeedham — APPROVED** [review]
Sounds like a useful change to me :)

**Comment by trevor-scheer** [comment]
@nikcork is it reasonable to say both? I think your comment brings some additional clarity to my description - that the `isRedirect` is not canonical.

Another good clarification:
>upon which you're basing your own interface

I shouldn't have said my own fetcher interface, rather I want to allow users to bring a fetcher which implements this `Fetch` like so:

```ts
interface Opts {
  // ...
  fetcher: Fetch;
}
```

Let me know if I'm missing any part of your question, I think it's a good one to ask! Makes me wonder if you're aware of a canonical fetch type that I'm not privy to, or maybe there ought to be one independent of the `node-fetch` typings instead?

**Comment by nikcorg** [comment]
@trevor-scheer thanks for clarifying.

While I agree that updating the node-fetch types so that they match a canonical type definition for the fetch API is worthwhile in the case that they are in their current form in conflict, updating the types so that they can be used in place of that interface is in my opinion going the wrong way about it.

Firstly, the @types/node-fetch package is only relevant for "legacy" versions of the node-fetch library; versions >= 3 [exports their own types](https://github.com/node-fetch/node-fetch#typescript). The package is thus subject for eventual removal. (They [were already once removed](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/55533), and the current incarnation was [reinstated to fix an error](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58693) in the types.) Existing versions of the package will of course continue to exist in the npm package repository, but is depending on a library whose deprecation is probably not too far off really something you want?

Secondly, node.js recently added [native fetch](https://nodejs.org/en/blog/release/v17.5.0/#add-fetch-api) into core (see e.g. https://news.ycombinator.com/item?id=30161626 and https://fusebit.io/blog/node-fetch/), meaning you could use `typeof global.fetch` in your interface. That does however come with caveats on node.js version requirements.

In summary, unless there is a canonical type definition that you can rely on, e.g. as part of @types/node, I would personally implement the minimal interface in the project that is accepting a fetch implementation, rather than relying on a third party library's types.

This is of course only my personal opinion, which is quite likely strongly influenced by the Go developer in me. (In Go, interfaces defined where a dependency is used is considered good practice.)

If you still feel like this PR adds value to the node-fetch types, I shan't stand in your way and will be more than happy to remove my roadblock. Sorry for being a bit of a stick in the mud.

**Comment by trevor-scheer** [comment]
@nikcorg no worries, I think it's good perspective. In hindsight, exporting the type here was probably more of an easy way out than it was a correct approach.

>  I would personally implement the minimal interface in the project that is accepting a fetch implementation
We actually went forward with exactly this and are pretty happy with the result.

In an ideal world, the undici / global fetch solution is preferable, but supporting LTS versions of node is a requirement for our projects. Some day though :)
